src/peft/tuners/xlora/layer.py

```diff
@@ -0,0 +1,186 @@
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,  # XLoraModel
+        target: lora.LoraLayer,
+        target_forward: Callable[..., Any],
+        layer_number: int,
+        config: XLoraConfig,
+    ) -> None:
+        self.model = model
+        self.target_forward = target_forward
+        self.target = target
+        self.layer_number = layer_number
+        self.config = config
+
+    @staticmethod
+    def apply_scalings_to_x(x: torch.Tensor, scalings_layer: torch.Tensor, adapter: int) -> torch.Tensor:
+        # scalings_layer = [batch_size, seq_len, n_classes]
+        scalings = scalings_layer[:, :, adapter].unsqueeze(-1)
+        # scalings_layer = [batch_size, seq_len, 1]
+        return x * scalings
+
+    def get_maybe_topk_scalings(self) -> torch.Tensor:
+        # xlora_scalings = [batch_size, seq_len, n_classes]
+        xlora_scalings: Tensor = self.model.internal_xlora_scalings[:, :, self.layer_number, :]  # type: ignore
+
+        if self.config.top_k_lora is not None:
+            _, topk_indices = torch.topk(xlora_scalings, k=self.config.top_k_lora, dim=-1)
+
+            # Mask the topk to True, the rest to False
+            mask = torch.zeros_like(xlora_scalings, dtype=torch.bool)
+            mask.scatter_(-1, topk_indices, True)
+
+            xlora_scalings = xlora_scalings * mask.to(xlora_scalings.dtype)
+
+        if self.config.enable_softmax_topk:
+            nonzero_mask = xlora_scalings != 0
+            softmax_res_nonzero = torch.softmax(xlora_scalings[nonzero_mask], dim=-1)
+            xlora_scalings[nonzero_mask] = softmax_res_nonzero
+
+        return xlora_scalings
+
+
+class XLoRALinearLayer(XLoRALayer):
+    def __init__(
+        self,
+        model: nn.Module,
+        target: lora.Linear,
+        target_forward: Callable[..., Any],
+        layer_number: int,
+        config: XLoraConfig,
+    ) -> None:
+        super().__init__(model, target, target_forward, layer_number, config)
+
+    def forward(self, x: Tensor, *args: Any, **kwargs: Any) -> Tensor:
+        """
+        This method is designed to be a drop-in-replacement for the LoRA layers' .forward method.
+        To use it, a bound method must be created (bound to an instance of the XLoRALayer class).
+        """
+
+        previous_dtype = x.dtype
+        xlora_scalings = self.get_maybe_topk_scalings()
+
+        if self.target.disable_adapters:
+            if self.target.merged:
+                self.target.unmerge()
+            result = self.target.base_layer(x, *args, **kwargs)
+        elif self.target.merged:
+            result = self.target.base_layer(x, *args, **kwargs)
+        else:
+            result = self.target.base_layer(x, *args, **kwargs)
+
+            for adapter_n, active_adapter in enumerate(self.target.active_adapters):
+                if active_adapter not in self.target.lora_A.keys():
+                    continue
+                lora_A = self.target.lora_A[active_adapter]
```

Comment:
Small caveat, if the target `LoraLayer` uses `DoRA`, the calculations below are not correct, as they don't take DoRA into account. I think the easiest thing to do is to check this like so:

```python
if target.use_dora[activate_adapter]:
    raise ValueError("X-LoRA currently does not support LoRA layers with DoRA")
```

The same logic would apply to the other layer types.

We could certainly also try to make DoRA work, it's up to you, but I'd be fine with excluding that use case for now.

Reply:
I have added that error check. It should be pretty easy to add that in the future though.